Introduce SdJwtFacade layer for simplified SD-JWT handling and enhance test coverage

Awambeng opened the pull request:
This PR introduces the SdJwtFacade as a new layer on top of the existing SD-JWT implementation. The new facade simplifies SD-JWT creation, verification, and string representation.

Additionally, it enhances test coverage with:

- Tests for SD-JWT creation, verification with valid/invalid keys, and string conversion.
- New test classes for Elytron and DefaultCrypto environments, including a check for FIPS mode.

Closes https://github.com/keycloak/keycloak/issues/32955

Awambeng committed:
feat(sdjwt): simplify SD-JWT creation and management, add comprehensive tests

Refactored SD-JWT creation to simplify the API with SdJwtFacade. Added tests for SD-JWT creation, validation with valid and invalid keys, string conversion, and integration with Elytron and DefaultCrypto.

Closes issue-32955

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

mposolda commented:
@tnorimat @francis-pouatcha @Captain-P-Goldfish Anyone of you available to review this PR please?

tnorimat commented:
@mposolda Hello,  I will review the PR.

tnorimat commented on core/src/main/java/org/keycloak/sdjwt/SdJwtFacade.java:
```diff
@@ -0,0 +1,71 @@
+package org.keycloak.sdjwt;
```
Could you add the follwoing copyright header?
```
/*
 * Copyright 2024 Red Hat, Inc. and/or its affiliates
 * and other contributors as indicated by the @author tags.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```

tnorimat commented on core/src/test/java/org/keycloak/sdjwt/SdJwtFacadeTest.java:
```diff
@@ -0,0 +1,126 @@
+package org.keycloak.sdjwt;
```
Could you add the follwoing copyright header?
```
/*
 * Copyright 2024 Red Hat, Inc. and/or its affiliates
 * and other contributors as indicated by the @author tags.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```

tnorimat commented on crypto/default/src/test/java/org/keycloak/crypto/def/test/sdjwt/DefaultCryptoSdJwtFacadeTest.java:
```diff
@@ -0,0 +1,18 @@
+package org.keycloak.crypto.def.test.sdjwt;
```
Could you add the follwoing copyright header?
```
/*
 * Copyright 2024 Red Hat, Inc. and/or its affiliates
 * and other contributors as indicated by the @author tags.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```

tnorimat commented on crypto/elytron/src/test/java/org/keycloak/crypto/elytron/test/sdjwt/ElytronCryptoSdJwtFacadeTest.java:
```diff
@@ -0,0 +1,9 @@
+package org.keycloak.crypto.elytron.test.sdjwt;
```
Could you add the follwoing copyright header?
```
/*
 * Copyright 2024 Red Hat, Inc. and/or its affiliates
 * and other contributors as indicated by the @author tags.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```

tnorimat commented on core/src/test/java/org/keycloak/sdjwt/SdJwtFacadeTest.java:
```diff
@@ -0,0 +1,126 @@
+
+        String sdJwtString = sdJwtFacade.getSdJwtString(sdJwt);
+
+        assertNotNull(sdJwtString);
```
It migt be better to compare `sdJwt.toString()` and `sJwtString` and confirm that the both are exactly matched.

tnorimat reviewed (commented):
@Awambeng Thank you for the PR. I added some review comments. Could you check them?

Awambeng commented:
> @Awambeng Thank you for the PR. I added some review comments. Could you check them?

@tnorimat  Thank you for your review and helpful feedback! I’ll address the comments as requested, including the copyright header and test adjustments.

Awambeng committed:
test: refine sdjwt string comparison for accuracy

Added copyright header and adjusted test to directly compare sdJwt.toString() with sdJwtString for exact match, enhancing test precision

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

tnorimat reviewed (approved):
LGTM

tnorimat commented:
@mposolda I approved the PR. Could you check the PR?

mposolda commented:
> @mposolda i am trying not to review tickets that are from adorsys. To have other eyes on the tickets.

@francis-pouatcha Sure, I understand. Thanks

mposolda reviewed (approved):
@Awambeng @tnorimat Thanks!

Awambeng commented:
> @Awambeng @tnorimat Thanks!

@mposolda Thank you too
